### ISSUE TITLE
Possible race condition or premature finishing workflows in Orquesta workflow engine

### ISSUE THREAD

**Issue by igcherkaev** [StackStorm/st2#5031]
## SUMMARY

Seems like there's some sort of race condition in orquesta where it would abruptly finish workflow.

### STACKSTORM VERSION

```
st2 3.2.0, on Python 2.7.5
```

##### OS, environment, install method

CentOS 7 following official docs with single st2workflowengine instance running and coordination with redis configured.

### Meta
```yaml
---
pack: "playground"
name: "wf_orquesta_test7"
description: "Orquesta test"
runner_type: orquesta
enabled: true
entry_point: "workflows/wf_orquesta_test7.yaml"
```

### Workflow

```yaml
---
version: '1.0'

vars:
  - notifications: true

tasks:
  where_all_begins:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - task_0
          - task_1
          - task_2
          - task_3

  task_0:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_0_ok_notification

  task_0_ok_notification:
    action: core.noop

  task_1:
    action: core.local
    input:
      cmd: "exit 1"
    next:
      - when: <% succeeded() %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_1_ok_notification
      - when: <% failed() %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_1_failed_notification

  task_1_failed_notification:
    action: core.noop

  task_1_ok_notification:
    action: core.noop

  task_2:
    action: core.local
    input:
      cmd: "exit 0"
    next:
      - when: <% succeeded() %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_2_ok_notification
      - when: <% failed() %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_2_failed_notification

  task_2_failed_notification:
    action: core.noop

  task_2_ok_notification:
    action: core.noop

  task_3:
    action: core.local
    input:
      cmd: "exit 1"
    next:
      - when: <% succeeded() %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_3_ok_notification
      - when: <% failed() %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_3_failed_notification

  task_3_failed_notification:
    action: core.noop

  task_3_ok_notification:
    action: core.noop

  where_all_ends_good:
    join: all
    action: core.noop

  where_all_ends_bad:
    # so far the only workaround I found is to have this "join: all" here, however, it's not always possible or wanted.
    # join: all
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - fail

```

## Expected Results

The workflow above has everything hard-coded and outcome must be same every run. However, in reality, results differ every run. Expected results is to see the following tasks executed before the workflow finishes:
- task_0_ok_notification
- task_1_failed_notification
- task_2_ok_notification
- task_3_failed_notification

## Actual Results

In reality I get different results every run:

![image](https://user-images.githubusercontent.com/18666442/91774495-8bc4f880-ebae-11ea-8926-4f20161dd96b.png)

![image](https://user-images.githubusercontent.com/18666442/91774527-9c756e80-ebae-11ea-86b0-89d01be20b32.png)

Notice that random notification tasks are missing. Sometimes they are all there, but most of the times one or some are missing.

**Comment by igcherkaev** [comment]
@m4dcoder is it possible that orquesta stops whole workflow execution as soon as one route finishes all tasks and overall result is FAILED regardless of other routes may be still running? Because it sure seems like so...

**Comment by m4dcoder** [comment]
Basically, the issue here is that you have to explicitly distinguish between `<% failed() and not ctx().notifications %>` and `<% failed() and ctx().notifications %>`. 

Let's take task1 as an example. In your original version, both transition `where_all_ends_bad ` and `task_1_failed_notification` are satisfied.  Since there is a `fail` at `where_all_ends_bad`, the workflow will fail fast and end after all tasks that have started are complete.  

In the version below, I modified 

```
      - when: <% failed() %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_1_failed_notification
```

to 

```
      - when: <% failed() and not ctx().notifications %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_1_failed_notification
```

Now, the engine knows to only execute one of these depending on what the value of `ctx().notifications` is.

Below is the version of the workflow that runs the tasks that you expected. 

```
---
version: '1.0'

vars:
  - notifications: true

tasks:
  where_all_begins:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - task_0
          - task_1
          - task_2
          - task_3

  task_0:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_0_ok_notification

  task_0_ok_notification:
    action: core.noop

  task_1:
    action: core.local
    input:
      cmd: "exit 1"
    next:
      - when: <% succeeded() and not ctx().notifications %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_1_ok_notification
      - when: <% failed() and not ctx().notifications %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_1_failed_notification

  task_1_failed_notification:
    action: core.noop

  task_1_ok_notification:
    action: core.noop

  task_2:
    action: core.local
    input:
      cmd: "exit 0"
    next:
      - when: <% succeeded() and not ctx().notifications %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_2_ok_notification
      - when: <% failed() and not ctx().notifications %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_2_failed_notification

  task_2_failed_notification:
    action: core.noop

  task_2_ok_notification:
    action: core.noop

  task_3:
    action: core.local
    input:
      cmd: "exit 1"
    next:
      - when: <% succeeded() and not ctx().notifications %>
        do:
          - where_all_ends_good
      - when: <% succeeded() and ctx().notifications %>
        do:
          - task_3_ok_notification
      - when: <% failed() and not ctx().notifications %>
        do:
          - where_all_ends_bad
      - when: <% failed() and ctx().notifications %>
        do:
          - task_3_failed_notification

  task_3_failed_notification:
    action: core.noop

  task_3_ok_notification:
    action: core.noop

  where_all_ends_good:
    join: all
    action: core.noop

  where_all_ends_bad:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - fail
```

**Comment by emptywee** [comment]
@m4dcoder  Ok, so you are saying that random outcome is by design in the engine? I honestly doubt it was designed that way.

> It seems like you made an assumption that the engine only will execute 1 of the task transition. In actual, the engine will execute all the task transition where the when condition is satisfied. 

I did not make such assumption. In fact, I want both to be executed. `ctx().notification` is a flag to send out a notification about workflow progression. And we set it to `true` here. All variables and exit codes are fixed and every run the engine should produce exact same result. However, it does not.

> Since there is a fail at where_all_ends_bad, the workflow will signal the engine to stop running the workflow after all tasks that have started are complete

That's what the issue is about. When it does `- fail`, it doesn't care that there are still un-executed transitions from previous tasks.

Please, take a look at the `Actual Results` section of the issue. It depicts that the engine produces random results and sometimes it executes transitions, sometimes it doesn't. Try to take my workflow and execute it multiple times and see if you get same results every time.

**Comment by nmaludy** [comment]
@emptywee I'm going to try and summarize your thoughts around behavior, let me know if what i'm saying is right.

You are expecting that all branches continue executing until they are exhausted and all tasks have completed, rather than being terminated because one branch fails. 

Does this sound right?

**Comment by emptywee** [comment]
Yes, @nmaludy, that sounds about right. Orquesta documentation says the same (though, of course, it can be interpreted differently, if one wants to):

> When there are no more tasks to execute, the workflow is complete.

> When there is no more tasks identified to run next, the workflow is complete.

In my case here there are tasks to run, however, they randomly don't get executed.

**Comment by m4dcoder** [comment]
@emptywee 

Since task 0-3 in your workflow are all executed in parallel, the order of these task execution are not deterministic.  They'll be picked up by the scheduler and dispatch to action runner workers to be run independently of each other. These action runner workers may introduce delays. There is no guarantee that they will be executed in the order that you define in the workflow. 

When `fail` is triggered in the workflow, the engine will wait for existing tasks to complete and then terminate the workflow regardless if there are tasks identified to run next. As soon as one of the tasks in your workflow fails and transition to `where_all_ends_bad`, the workflow will terminate. This is by design.

As far as the original issue is concerned, there is no race here.If you want the `fail` to behave differently, please submit a separate request for feature enhancement.

**Comment by nmaludy** [comment]
@emptywee trying to figure out how to solve your problem given the constraints around the way Orquesta operates.

Wondering if you can create a state variable called `wf_failed` and set that variable to `true` rather than explicitly transitioning to `fail`. Instead we do something like:

```
- when: "{{ failed() }}"
  publish:
    - wf_failed: true
  next:
    - join_task
```

This will cause the branch to finish and then we can join all of the branches back together and check for `wf_failed`. Thoughts on this, @emptywee @m4dcoder ?

**Comment by emptywee** [comment]
I don't understand, @m4decoder, it's okay for the engine not to execute
transitions randomly just because one of the tasks transitions to "fail"?
Why is it different then for assumed "continue" transition?

Documentation says that orquesta will complete all tasks before finishing a
workflow. Mistral also would run all transitions that match the conditions
defined.

Maybe there's no race per se, but definitely it is not very intuitive - you
define transitions and conditions, and while they are met, the engine
decides to end workflow execution without executing them.

@nmaludy join indeed would help in this particular case, but it's not
always desirable or  possible to do. I'd rather have the engine complete
all branching and transitions before completing the workflow.

**Comment by emptywee** [comment]
Sorry, was typing my reply from the phone, meant @m4dcoder of course.

**Comment by emptywee** [comment]
I replaced the `where_all_ends_bad` task from being a `core.noop` + `fail` transition to a `core.local` action with `exit 1`:
```
  where_all_ends_bad:
    # join: all
    action: core.local
    input:
      cmd: exit 1
```

Where's `task_2_ok_notification` task in the execution?

![image](https://user-images.githubusercontent.com/4997180/95134508-9b44dd80-0728-11eb-9ec3-28a9a9d4652f.png)

Or where's `task_2_ok_notification` and `task_1_failed_notification` here?

![image](https://user-images.githubusercontent.com/4997180/95134551-b1eb3480-0728-11eb-8639-d3b74e2bef3d.png)

There's no explicit fail now in the workflow, but still it doesn't execute all tasks that it must execute prior to finish it. I still insist it's a bug, @m4dcoder

**Comment by emptywee** [comment]
> Since task 0-3 in your workflow are all executed in parallel, the order of these task execution are not deterministic. They'll be picked up by the scheduler and dispatch to action runner workers to be run independently of each other. These action runner workers may introduce delays. There is no guarantee that they will be executed in the order that you define in the workflow

I don't expect it to run them in any particular order, so this is not a part of the issue here.

> When fail is triggered in the workflow, the engine will wait for existing tasks that are already running to complete and then terminate the workflow regardless if there are tasks identified to run next. As soon as one of the tasks in your workflow fails and transition to where_all_ends_bad, the workflow will terminate. This is by design.

Apparently not only by `fail`, but any other action which doesn't succeed. It's a very poor design if you don't get consistent results with fixed variables and outcomes of each task, in my opinion.

If we're to keep this design, it should be documented, that as soon as any branch of a given workflow fails, it fails the whole execution regardless of having other branches still have tasks to execute. This is a big and important thing.

**Comment by emptywee** [comment]
In other words, is this by design?

![image](https://user-images.githubusercontent.com/4997180/95137929-b581ba00-072e-11eb-971a-b0dd1a6e635e.png)

![image](https://user-images.githubusercontent.com/4997180/95137952-bdd9f500-072e-11eb-9a2e-8a3d53615c79.png)

```
---
version: '1.0'

vars:
  - notifications: true

tasks:
  where_all_begins:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - branch_a
          - branch_b

  branch_a:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - branch_a_task_1

  branch_a_task_1:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - branch_a_task_2

  branch_a_task_2:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - branch_a_task_3

  branch_a_task_3:
    action: core.noop
    next:
      - when: <% succeeded() %>
        do:
          - branch_a_task_4

  branch_a_task_4:
    action: core.noop

  branch_b:
    action: core.local
    input:
      cmd: "exit 1"
```

**Comment by m4dcoder** [comment]
Yes, currently any failure in the workflow in one branch or another will terminate the workflow. Please open an enhancement request with proposed changes, we will see what we can do. As far as this issue is concerned, this is not a bug/issue regardless if you think this is a poor design or not.  If you think this information is not documented properly, we can address that.

**Comment by emptywee** [comment]
Mistral version of the same workflow provides consistent results, so for the folks who migrates from mistral it's a real pain :(

![image](https://user-images.githubusercontent.com/4997180/95138471-e9a9aa80-072f-11eb-96e6-7f13c6d4cc9e.png)

@m4dcoder mmm, okay, we probably have different interpretation of the documentation, let's assume it's my poor English that I misinterpreted it. I'll open another issue to discuss the current design for failures.

**Comment by amanda11** [comment]
I had a good read of the documentation, and I think the behaviour is implied, but I couldn't find a place where it was explicitly pointed out. If we added a statement similar to "The workflow execution will stop once a workflow branch fails, or all branches are complete. If a workflow branch fails then all current in-progress tasks would be completed, but no further tasks on those branches would be run"  would that match what we have and make it clearer?

**Comment by emptywee** [comment]
@amanda11 it would be good to know that, especially for those who migrate from mistral to orquesta, because behavior of the two engines is quite different. And one would expect them to behave similarly, but, alas, I stepped on these rakes and you can judge by the volume of my screaming it's very painful :(

**Comment by amanda11** [comment]
I will try and get a doc Pr up for review today.
